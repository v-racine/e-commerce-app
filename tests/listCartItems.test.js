const AppFactory = require('../src/app');
const request = require('supertest');

describe('list cart items', () => {
  const mockCartsRepo = {};
  const mockProductsRepo = {};

  const app = AppFactory({
    cartsRepo: mockCartsRepo,
    productsRepo: mockProductsRepo,
  });

  describe('when: there is no session cart id and the user requests to view their cart', () => {
    let rsp;

    beforeEach(async () => {
      rsp = await request(app).get('/cart').send();
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    test('then: we redirect to home', async () => {
      const text = rsp.text;
      expect(text).toMatchSnapshot();

      const status = rsp.status;
      expect(status).toBe(302);
    });
  });

  describe('when: there is a session cart id and the user requests to view their cart', () => {
    let rsp;

    const cartId = 'sdlkfhs';

    const products = [
      {
        id: 'dflkhs',
        quantity: 1,
      },
      {
        id: '3432423',
        quantity: 3222,
      },
    ];

    beforeEach(async () => {
      mockCartsRepo.create = jest.fn().mockImplementation(() => {
        return new Promise((resolve) => {
          resolve({ id: cartId, items: [] });
        });
      });

      mockCartsRepo.getOne = jest.fn().mockImplementation(() => {
        return new Promise((resolve) => {
          resolve({ id: cartId, items: products });
        });
      });

      mockProductsRepo.getOne = jest.fn().mockImplementation((itemId) => {
        return new Promise((resolve) => {
          for (const product of products) {
            if (product.id === itemId) {
              resolve({ id: itemId, title: `product_${itemId}`, price: 12 });
            }
          }
        });
      });

      const createRsp = await request(app).post('/cart').send();

      const cookie = createRsp.get('Set-Cookie');

      rsp = await request(app).get('/cart').set('Cookie', cookie).send();
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    test('then: call cartsRepo.getOne, we call productsRepo.getOne for each item, and display the cart items', async () => {
      expect(mockCartsRepo.getOne).toHaveBeenCalledWith(cartId);

      expect(mockProductsRepo.getOne).toHaveBeenCalledTimes(products.length);

      for (let callCount = 1; callCount <= products.length; callCount++) {
        expect(mockProductsRepo.getOne).toHaveBeenNthCalledWith(
          callCount,
          products[callCount - 1].id,
        );
      }

      expect(rsp.text).toMatchSnapshot();
    });
  });

  describe('when: there is a session cart id and the user requests to view their cart but the cart is empty', () => {
    let rsp;

    const cartId = 'papillon';

    const products = [];

    beforeEach(async () => {
      mockCartsRepo.create = jest.fn().mockImplementation(() => {
        return new Promise((resolve) => {
          resolve({ id: cartId, items: [] });
        });
      });

      mockCartsRepo.getOne = jest.fn().mockImplementation(() => {
        return new Promise((resolve) => {
          resolve({ id: cartId, items: products });
        });
      });

      mockProductsRepo.getOne = jest.fn();

      const createRsp = await request(app).post('/cart').send();

      const cookie = createRsp.get('Set-Cookie');

      rsp = await request(app).get('/cart').set('Cookie', cookie).send();
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    test('then: call cartsRepo.getOne, we do not call productsRepo.getOne at all, and display some html to the user', async () => {
      expect(mockCartsRepo.getOne).toHaveBeenCalledWith(cartId);

      expect(mockProductsRepo.getOne).toHaveBeenCalledTimes(0);

      expect(rsp.text).toMatchSnapshot();
      expect(rsp.status).toBe(200);
    });
  });
});
