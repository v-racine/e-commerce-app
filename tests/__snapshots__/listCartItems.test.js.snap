// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`list cart items when: there is a session cart id and the user requests to view their cart but the cart is empty then: call cartsRepo.getOne, we do not call productsRepo.getOne at all, and display some html to the user 1`] = `
"
    <!DOCTYPE html>
      <html lang="en">
      <head>
        <meta charset="utf-8">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        <title>Shop</title>
        
        <link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/all.min.css" rel="stylesheet">
        <link href="/css/main.css" rel="stylesheet">
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/bulma/0.7.5/css/bulma.min.css"></link>
      </head>

      <body>
        <header>
          <nav class="navbar navbar-top">
            <div class="container navbar-container">
              <div>
                <ul class="social">
                  <li>
                    <a href=""><i class="fa fa-phone"></i>+1 555 987 6543</a>
                  </li>
                  <li>
                    <a href=""><i class="fa fa-envelope"></i> shop@myshop.com</a>
                  </li>
                </ul>
              </div>
              <div>
                <ul class="social">
                  <li><a href=""><i class="fab fa-facebook"></i></a></li>
                  <li><a href=""><i class="fab fa-twitter"></i></a></li>
                  <li><a href=""><i class="fab fa-linkedin"></i></a></li>
                  <li><a href=""><i class="fab fa-dribbble"></i></a></li>
                  <li><a href=""><i class="fab fa-google-plus"></i></a></li>
                </ul>
              </div>
            </div>
          </nav>
          <nav class="navbar navbar-bottom">
            <div class="container navbar-container">
              <div>
                <a href="/">
                  <h3 class="title">EComm Shop</h3>
                </a>
              </div>
              <div class="navbar-item">
                <div class="navbar-buttons">
                  <div class="navbar-item">
                    <a href="/"><i class="fa fa-star"></i> Products</a>
                  </div>
                  <div class="navbar-item">
                    <a href="/cart"><i class="fa fa-shopping-cart"></i> Cart</a>
                  </div>
                </div>
              </div>
            </div>
          </nav>
        </header>

        
      <div id="cart" class="container">
        <div class="columns">
          <div class="column"></div>
          <div class="column is-four-fifths">
            <h3 class="subtitle"><b>Shopping Cart</b></h3>
            <div>
              
            </div>
            <div class="total message is-info">
              <div class="message-header">
                Total
              </div>
              <h1 class="title">$0</h1>
              <button class="button is-primary">Buy</button>
            </div>
          </div>
          <div class="column"></div>
        </div>
      </div>
    
      </body>
    </html>
  "
`;

exports[`list cart items when: there is a session cart id and the user requests to view their cart then: call cartsRepo.getOne, we call productsRepo.getOne for each item, and display the cart items 1`] = `
"
    <!DOCTYPE html>
      <html lang="en">
      <head>
        <meta charset="utf-8">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        <title>Shop</title>
        
        <link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/all.min.css" rel="stylesheet">
        <link href="/css/main.css" rel="stylesheet">
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/bulma/0.7.5/css/bulma.min.css"></link>
      </head>

      <body>
        <header>
          <nav class="navbar navbar-top">
            <div class="container navbar-container">
              <div>
                <ul class="social">
                  <li>
                    <a href=""><i class="fa fa-phone"></i>+1 555 987 6543</a>
                  </li>
                  <li>
                    <a href=""><i class="fa fa-envelope"></i> shop@myshop.com</a>
                  </li>
                </ul>
              </div>
              <div>
                <ul class="social">
                  <li><a href=""><i class="fab fa-facebook"></i></a></li>
                  <li><a href=""><i class="fab fa-twitter"></i></a></li>
                  <li><a href=""><i class="fab fa-linkedin"></i></a></li>
                  <li><a href=""><i class="fab fa-dribbble"></i></a></li>
                  <li><a href=""><i class="fab fa-google-plus"></i></a></li>
                </ul>
              </div>
            </div>
          </nav>
          <nav class="navbar navbar-bottom">
            <div class="container navbar-container">
              <div>
                <a href="/">
                  <h3 class="title">EComm Shop</h3>
                </a>
              </div>
              <div class="navbar-item">
                <div class="navbar-buttons">
                  <div class="navbar-item">
                    <a href="/"><i class="fa fa-star"></i> Products</a>
                  </div>
                  <div class="navbar-item">
                    <a href="/cart"><i class="fa fa-shopping-cart"></i> Cart</a>
                  </div>
                </div>
              </div>
            </div>
          </nav>
        </header>

        
      <div id="cart" class="container">
        <div class="columns">
          <div class="column"></div>
          <div class="column is-four-fifths">
            <h3 class="subtitle"><b>Shopping Cart</b></h3>
            <div>
              
        <div class="cart-item message">
          <h3 class="subtitle">product_dflkhs</h3>
          <div class="cart-right">
            <div>
              $12  X  1 = 
            </div>
            <div class="price is-size-4">
              $12
            </div>
            <div class="remove">
              <form method="POST" action="/cart/products/delete">
                <input hidden value="dflkhs" name="itemId" /> 
                <button class="button is-danger">                  
                  <span class="icon is-small">
                    <i class="fas fa-times"></i>
                  </span>
                </button>
              </form>
            </div>
          </div>
        </div>
      
        <div class="cart-item message">
          <h3 class="subtitle">product_3432423</h3>
          <div class="cart-right">
            <div>
              $12  X  3222 = 
            </div>
            <div class="price is-size-4">
              $38664
            </div>
            <div class="remove">
              <form method="POST" action="/cart/products/delete">
                <input hidden value="3432423" name="itemId" /> 
                <button class="button is-danger">                  
                  <span class="icon is-small">
                    <i class="fas fa-times"></i>
                  </span>
                </button>
              </form>
            </div>
          </div>
        </div>
      
            </div>
            <div class="total message is-info">
              <div class="message-header">
                Total
              </div>
              <h1 class="title">$38676</h1>
              <button class="button is-primary">Buy</button>
            </div>
          </div>
          <div class="column"></div>
        </div>
      </div>
    
      </body>
    </html>
  "
`;

exports[`list cart items when: there is no session cart id and the user requests to view their cart then: we redirect to home 1`] = `"Found. Redirecting to /"`;
